runtime/lua/vim/treesitter/query.lua

```diff
@@ -184,6 +184,76 @@ local function read_query_files(filenames)
   return table.concat(contents, '')
 end
 
+---@param node TSNode
+---@param bufnr integer
+---@return integer, integer, integer, integer
+local function node_trim_trailing_blank_lines(node, bufnr)
+  local start_row, start_col, end_row, end_col = node:range()
+
+  -- Don't trim if region ends in middle of a line
+  if end_col == 0 then
+    while end_row >= start_row do
+      -- As we only care when end_col == 0, always inspect one line above end_row.
+      local end_line = api.nvim_buf_get_lines(bufnr, end_row - 1, end_row, true)[1]
+
+      if end_line ~= '' then
+        break
+      end
+
+      end_row = end_row - 1
+    end
+  end
+
+  return start_row, start_col, end_row, end_col
+end
+
+---@param node TSNode
+---@param bufnr integer
+---@return integer, integer, integer, integer
+local function node_trim_all_whitespace(node, bufnr)
```

Comment:
can we generalize this? would using `vim.fn.trim()` reduce the code?

```suggestion
---@param lines_only boolean
local function node_trim(node, bufnr, lines_only)
```

Reply:
Hmm, how do you think this should work? I can see the trimming itself being helpful but the main logic we need is the "trimming" of the range values, and I don't know how `vim(.fn).trim` would be able to help with that

Reply:
As for the generalization, I did that initially but the two code paths are pretty different, I couldn't really reuse much code so to me it just felt like separating them completely was easier on the eyes

Reply:
Keep in mind vim.fn.* functions that work with strings have problems with the vimscript bridge and are less efficient since the string gets copied multiple times. Both of those could be a problem here.

Reply:
I'm just surprised this requires so much code. Can it be simplified

Reply:
Ideally yeah... I tried some variations but sadly one reason it's difficult is the old behavior doesn't even trim whitespace lines, specifically empty (zero text on them) lines. So I couldn't even reuse one of the conditional branches in the "full trim" code. I think the only other generalized code that could be abstracted (since it is shared between the two) is `local start_row, start_col, end_row, end_col = node:range()`, but one line seems trivial at that point

Reply:
It would be worth checking if that difference is meaningful -- what would break if whitespace lines are trimmed as well? The `#trim!` directive is only used in (four, _not_ including Python ;)) folding queries, so arguably that would even be an improvement.